azure-ts-dynamicresource/index.ts

```diff
@@ -0,0 +1,86 @@
+import * as pulumi from "@pulumi/pulumi";
+import * as azure from "@pulumi/azure";
+
+import { CDNCustomDomainResource } from "./cdnCustomDomain";
+
+const location = "West US";
+
+/** The custom domain host of the CDN endpoint. */
+export let cdnCustomDomainResource: CDNCustomDomainResource | undefined;
+
+// Create an Azure Resource Group
+const resourceGroup = new azure.core.ResourceGroup("resourceGroup", {
+    location: location,
+});
+
+// Create an Azure resource (Storage Account) so we can point the CDN endpoint to it.
+const storageAccount = new azure.storage.Account("storage-account", {
+    resourceGroupName: resourceGroup.name,
+    location: resourceGroup.location,
+    accountTier: "Standard",
+    accountReplicationType: "LRS",
+});
+
+// Create a Blob container in the storage account.
+const blobContainer = new azure.storage.Container("blob-container", {
+    resourceGroupName: resourceGroup.name,
+    storageAccountName: storageAccount.name,
+    containerAccessType: "blob"
+});
+
+const cdnProfile = new azure.cdn.Profile("cdn-profile", {
+    location: location,
+    resourceGroupName: resourceGroup.name,
+    sku: "Standard_Akamai",
+});
+
+const cdnEndpoint = new azure.cdn.Endpoint("my-cdn-endpoint", {
+    name: "my-cdn-endpoint",
+    resourceGroupName: resourceGroup.name,
+    profileName: cdnProfile.name,
+    isHttpsAllowed: true,
+    isHttpAllowed: false,
+    location: location,
+    isCompressionEnabled: true,
+    originHostHeader: storageAccount.primaryBlobEndpoint,
+    contentTypesToCompresses: [
+        "text/plain",
+        "text/html",
+        "text/css",
+        "text/javascript",
+        "application/x-javascript",
+        "application/javascript",
+        "application/json",
+        "application/xml",
+        "image/png",
+        "image/jpeg"
+    ],
+    origins: [
+        {
+            name: "cdn-origin",
+            hostName: storageAccount.primaryBlobEndpoint,
+            httpsPort: 443,
+        }
+    ]
+});
+
+export const cdnEndpointUrl = pulumi.interpolate `https://${cdnEndpoint.hostName}`;
+
+pulumi.all([resourceGroup.name, cdnProfile.name, cdnEndpoint.name])
```

Comment:
I think it would look more natural if we accepted `Input<T>` in `CustomDomainOptions` and then on the client side it would be just a constructor call without `pulumi.all`.

Reply:
Yes, I definitely agree. In fact, that is what I started with. But if you look closely at the formal parameters of the functions in my dynamic resource provider, I used the type `CustomDomainOptions` for `inputs` instead of `any`. That means I will need to deal with an `Input<T>` inside those functions, which I couldn't see a way to unwrap them there without returning a `pulumi.Output<T>`, which I couldn't find a cleaner way to transform. Things got really confusing there to be honest.

I see that in your static website example, you are retaining the type of the `inputs` formal param as `any` and are accepting an `Input<T>` for your provider input. I am not sure how when your `create()` is called, that you are getting a regular string instead of an `Input<T>`. In any case, I would prefer to have the `inputs` param typed concretely instead of having to treat it as an anonymous object.

Reply:
Great point! I stitched my example from several other implementations, and the reason behind `any` is exactly as you say. Maybe @jen20 can clarify how `Input`s get resolved and where we stand on using types everywhere?

Reply:
Yeah I wonder if I could do a union type here for my args like: `pulumi.Input<T> | string`, and then still be able to pass an `Input<T>` from the outside, but within my dynamic resource use them as regular strings?

Reply:
So what we can do is to create two distinct types: an internal `Props` (name not final) which has no inputs in it and an exported `Args` which has the same shape but has all inputs. `Props` will be used in the dynamic provider implementation and `Args` is like normal args in all our components - to be passed into the constructor.

In order to avoid copy-pasting and typos, we can define a mapped type. I tried it on my static website example:

``` ts
// A mapped type to convert any interface into a type of the same shape but all types converted from T to Input<T>
type Inputs<T> = {
    [P in keyof T]: pulumi.Input<T[P]>;
}
  
interface StorageStaticWebsiteProps {
    accountName: string;
}

export type StorageStaticWebsiteArgs = Inputs<StorageStaticWebsiteProps>;
```

I was able to replace all `any` with `StorageStaticWebsiteProps`. Works like a charm.

@CyrusNajmabadi @lukehoban Don't we have a type like `Inputs` above in the standard Pulumi library? Do we what one?

Reply:
This is an elegant solution and one that I prefer, but would actually love it if it was part of the dynamic providers library itself. Aside from that, I do feel mapped types are sort of an advanced concept in TS for most beginners who are using TS, let alone Pulumi. That's just my opinion, though. I'd love to see what @lukehoban thinks about this.